kernel/src/sched/cooperative.rs

```diff
@@ -0,0 +1,170 @@
+//! Cooperative Scheduler for Tock
+//! When hardware interrupts occur while a userspace process is executing,
+//! this scheduler executes the top half of the interrupt,
+//! and then stops executing the userspace process immediately and handles the bottom
+//! half of the interrupt. However it then continues executing the same userspace process
+//! that was executing.
+
+use crate::callback::AppId;
+use crate::capabilities;
+use crate::common::dynamic_deferred_call::DynamicDeferredCall;
+use crate::common::list::{List, ListLink, ListNode};
+use crate::ipc;
+use crate::platform::mpu::MPU;
+use crate::platform::{Chip, Platform};
+use crate::process;
+use crate::sched::{Kernel, Scheduler};
+use crate::syscall::{ContextSwitchReason, Syscall};
+
+/// A node in the linked list the scheduler uses to track processes
+pub struct CoopProcessNode<'a> {
+    appid: AppId,
+    next: ListLink<'a, CoopProcessNode<'a>>,
+}
+
+impl<'a> CoopProcessNode<'a> {
+    pub fn new(appid: AppId) -> CoopProcessNode<'a> {
+        CoopProcessNode {
+            appid: appid,
+            next: ListLink::empty(),
+        }
+    }
+}
+
+impl<'a> ListNode<'a, CoopProcessNode<'a>> for CoopProcessNode<'a> {
+    fn next(&'a self) -> &'a ListLink<'a, CoopProcessNode> {
+        &self.next
+    }
+}
+
+/// Cooperative Scheduler
+pub struct CooperativeSched<'a> {
+    kernel: &'static Kernel,
+    pub processes: List<'a, CoopProcessNode<'a>>,
+}
+
+impl<'a> CooperativeSched<'a> {
+    pub const fn new(kernel: &'static Kernel) -> CooperativeSched<'a> {
+        CooperativeSched {
+            kernel: kernel,
+            processes: List::new(),
+        }
+    }
+
+    /// Executes a process until it yields or is interrupted. Returns whether it was interrupted.
+    unsafe fn do_process<P: Platform, C: Chip>(
+        &self,
+        platform: &P,
+        chip: &C,
+        process: &dyn process::ProcessType,
+        ipc: Option<&crate::ipc::IPC>,
+    ) -> bool {
+        loop {
+            if chip.has_pending_interrupts() {
+                break;
+            }
+
+            match process.get_state() {
+                process::State::Running => {
+                    // Running means that this process expects to be running,
+                    // so go ahead and set things up and switch to executing
+                    // the process.
+                    process.setup_mpu();
+                    chip.mpu().enable_mpu();
+                    let context_switch_reason = process.switch_to();
+                    chip.mpu().disable_mpu();
```

Comment:
This will need to be rebased on: https://github.com/tock/tock/pull/1822

Reply:
I think that #1822 changed such that I no longer have to rebase after all